crates/sui-types/src/event.rs

```diff
@@ -1,22 +1,127 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use move_core_types::language_storage::StructTag;
+use move_bytecode_utils::{layout::TypeLayoutBuilder, module_cache::GetModule};
+use move_core_types::{
+    language_storage::{ModuleId, StructTag, TypeTag},
+    value::{MoveStruct, MoveTypeLayout},
+};
+use name_variant::NamedVariant;
 use serde::{Deserialize, Serialize};
 use serde_with::{serde_as, Bytes};
+use strum_macros::EnumDiscriminants;
 
-/// User-defined event emitted by executing Move code.
-/// Executing a transaction produces an ordered log of these
+use crate::{
+    base_types::{ObjectID, SequenceNumber, SuiAddress, TransactionDigest},
+    committee::EpochId,
+    error::SuiError,
+    messages_checkpoint::CheckpointSequenceNumber,
+};
+
+/// A universal Sui event type encapsulating different types of events
+#[derive(Debug, Clone, PartialEq)]
+pub struct EventEnvelope {
+    /// UTC timestamp in milliseconds since epoch (1/1/1970)
+    timestamp: u64,
+    /// Transaction digest of associated transaction, if any
+    tx_digest: Option<TransactionDigest>,
+    /// Specific event type
+    event: SuiEvent,
+}
+
+impl EventEnvelope {
+    pub fn new(timestamp: u64, tx_digest: Option<TransactionDigest>, event: SuiEvent) -> Self {
+        Self {
+            timestamp,
+            tx_digest,
+            event,
+        }
+    }
+
+    pub fn event_type(&self) -> &'static str {
+        self.event.variant_name()
+    }
+}
+
+#[derive(Eq, Debug, Clone, PartialEq, Deserialize, Serialize, Hash)]
+pub enum TransferType {
+    Coin,
+    ToAddress,
+    ToObject, // wrap object in another object
+}
+
+/// Specific type of event
 #[serde_as]
-#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
-pub struct Event {
-    pub type_: StructTag,
-    #[serde_as(as = "Bytes")]
-    pub contents: Vec<u8>,
+#[derive(
+    Eq, Debug, Clone, PartialEq, NamedVariant, Deserialize, Serialize, Hash, EnumDiscriminants,
+)]
+#[strum_discriminants(name(SuiEventType))]
+pub enum SuiEvent {
```

Comment:
Why renaming Event to SuiEvent?

Just a naming nit but will be nice if we can keep using Event for internal types, as we have been using the `Sui` prefix for all the external type that are JSON/readability optimised. see https://github.com/MystenLabs/sui/blob/4ebe1a7aa17a0d0787a0d899c4ec52fc2afe4314/crates/sui-core/src/gateway_types.rs

Reply:
@patrickkuo got it, will rename it back to event.  (It was originally named differently because it was wrapping an Event in a SuiEvent, but no longer needed)